src/transaction/duck_transaction.cpp

```diff
@@ -271,23 +271,20 @@ unique_ptr<StorageLockKey> DuckTransaction::TryGetCheckpointLock() {
 }
 
 shared_ptr<CheckpointLock> DuckTransaction::SharedLockTable(DataTableInfo &info) {
-	lock_guard<mutex> l(active_locks_lock);
-	auto entry = active_locks.find(info);
-	if (entry != active_locks.end()) {
-		// found an existing lock
-		auto lock_weak_ptr = entry->second;
-		// check if it is expired
-		auto lock = lock_weak_ptr.lock();
-		if (lock) {
-			// not expired - return it
-			return lock;
-		}
+	unique_lock<mutex> transaction_lock(active_locks_lock);
+	auto &active_table_lock = active_locks[info];
```

Comment:
I think this might not be safe. We're keeping a reference to an `ActiveTableLock`, which is an object that sits in an `unordered_map`. Other threads can then modify the `unordered_map` (and add new elements to it). My suspicion is that, if the unordered_map is resized, this can potentially blow up.

Can we turn `ActiveTableLock` into a `unique_ptr` to prevent this from happening?

Reply:
People talking about unordered_map! I think iterators can become invalid on resize, but reference to object in the std::unordered_map are always valid.

This is confirmed by:
> References to elements in the [unordered_map](https://cplusplus.com/unordered_map) container remain valid in all cases, even after a rehash.
at https://cplusplus.com/reference/unordered_map/unordered_map/insert/

This is one of the classic complaints on std::unordered_map, that basically always require an extra level of indirection (and so it's impossible to make them faster while keeping the same interface, and changing interface to not rely on element validity will break stuff in hidden ways).

That said, making this explicit might be better.

Reply:
I've reworked this in [43237a6](https://github.com/duckdb/duckdb/pull/14370/commits/43237a6c245fa5caa787a526e459158ea5131625).